sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,536 @@
+            success = true;
+            return indexInput;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(fc);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for writing
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.createOutput(name, context);
+        ensureOpen();
+        OutputStream fos = Files.newOutputStream(directory.resolve(name), StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+        Cipher cipher = CipherFactory.getCipher(provider);
+        SecureRandom random = Randomness.createSecure();
+        byte[] iv = new byte[CipherFactory.IV_ARRAY_LENGTH];
+        random.nextBytes(iv);
+        if (dataKey == null) throw new RuntimeException("dataKey is null!");
+        CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.ENCRYPT_MODE, 0);
+        ivMap.put(getDirectory() + "/" + name, Base64.getEncoder().encodeToString(iv));
+        return new CryptoIndexOutput(name, fos, cipher);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param prefix the desired temporary file prefix
+     * @param suffix the desired temporary file suffix
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
+        if (prefix.contains("segments_") || prefix.endsWith(".si")) return super.createTempOutput(prefix, suffix, context);
+        ensureOpen();
+        String name;
+        while (true) {
+            name = getTempFileName(prefix, suffix, nextTempFileCounter.getAndIncrement());
+            OutputStream fos = Files.newOutputStream(directory.resolve(name), StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+            Cipher cipher = CipherFactory.getCipher(provider);
+            SecureRandom random = Randomness.createSecure();
+            byte[] iv = new byte[CipherFactory.IV_ARRAY_LENGTH];
+            random.nextBytes(iv);
+            CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.ENCRYPT_MODE, 0);
+            ivMap.put(getDirectory() + "/" + name, Base64.getEncoder().encodeToString(iv));
+            return new CryptoIndexOutput(name, fos, cipher);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void close() throws IOException {
+        try {
+            deleteFile("ivMap");
+        } catch (java.nio.file.NoSuchFileException fnfe) {
+
+        }
+        IndexOutput out = super.createOutput("ivMap", new IOContext());
+        out.writeMapOfStrings(ivMap);
+        out.close();
+        isOpen = false;
+        deletePendingFiles();
```

Comment:
```deletePendingFiles``` is also synchronised.. can we get into a deadlock?

we need to use ReentrantLock lock here.